### PR TITLE
feat(staking): adds commission + commission_rate applied

### PR DESCRIPTION
**Pull request by damirka**
- commission_rate is set initially
- any change is applied in E+2 no difference if the node is in the committee or not
- changes `commission_rate` type to `u16`, since it's in BP
- slightly improves API of `PendingValues` allowing access to the inner `VecMap`
- adds `extra_fields` Bag to `StakingPool` to make it future proof

### CODE REVIEW

File: contracts/walrus/sources/staking/staking_pool.move [hunk 1]
```diff
@@ -95,9 +97,13 @@ public struct StakingPool has key, store {
     /// stakes. Token amount for these principals is calculated via the exchange
     /// rate at the activation epoch.
     pending_early_withdrawals: PendingValues,
+    /// The pending commission rate for the pool. Commission rate is applied in
+    /// E+2, so we store the value for the matching epoch and apply it in the
+    /// `advance_epoch` function.
+    pending_commission_rate: PendingValues,
     /// The commission rate for the pool.
     /// TODO: allow changing the commission rate in E+2.
```
Comment (karlwuest) [inline contracts/walrus/sources/staking/staking_pool.move]: Nit: Remove TODO. Also, let's specify in the comment that the commission rate is given as basis points.

File: contracts/walrus/sources/staking/staking_pool.move [hunk 2]
```diff
@@ -398,8 +426,13 @@ public(package) fun process_pending_stake(pool: &mut StakingPool, wctx: &WalrusC
 /// Sets the next commission rate for the pool.
 /// TODO: implement changing commission rate in E+2, the change should not be
 /// immediate.
```
Comment (karlwuest) [inline contracts/walrus/sources/staking/staking_pool.move]: Nit: Remove TODO

File: contracts/walrus/tests/e2e_tests.move [hunk 3]
```diff
@@ -8,7 +8,7 @@ use std::unit_test::assert_eq;
 use walrus::{e2e_runner, staking_pool, test_node, test_utils};
 use sui::test_scenario;
 
-const COMMISSION: u64 = 1;
+const COMMISSION: u16 = 0;
```
Comment (karlwuest) [inline contracts/walrus/tests/e2e_tests.move]: Nit
```suggestion
const COMMISSION_RATE: u16 = 0;
```

File: contracts/walrus/tests/staking/staking_pool/staking_pool_tests.move [hunk 4]
```diff
@@ -588,24 +588,25 @@ fun test_advance_pool_epoch() {
...
     assert_eq!(pool.storage_price(), 1);
     assert_eq!(pool.node_capacity(), 1);
 
-    // pool changes commission rate to 100 in epoch E+1
-    pool.set_next_commission(100);
     pool.set_next_node_capacity(1000);
     pool.set_next_storage_price(100);
     pool.set_next_write_price(100);
 
+    // pool changes commission rate to 1% in epoch E+2
```
Comment (karlwuest) [inline contracts/walrus/tests/staking/staking_pool/staking_pool_tests.move]:
```suggestion
    // pool changes commission rate to 10% in epoch E+2
```

File: crates/walrus-sui/src/types/move_structs.rs [hunk 5]
```diff
@@ -452,6 +465,15 @@ where
...
 
+#[instrument(err, skip_all)]
+fn deserialize_bag<'de, D>(deserializer: D) -> Result<ObjectID, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let (object_id, _length): (ObjectID, u64) = Deserialize::deserialize(deserializer)?;
+    Ok(object_id)
+}
+
```
Comment (karlwuest) [inline crates/walrus-sui/src/types/move_structs.rs]: Suggestion: Just rename `deserialize_table` to `deserialize_bag_or_table` and use the same function for both deserializations.

File: contracts/walrus/sources/staking/staking_pool.move (hunk 1 above)
Comment (damirka) [reply contracts/walrus/sources/staking/staking_pool.move]: Nice catch!

**Pull request merged**